### DISCUSSION TITLE
Rendering islands from Routes and best way to load data with islands feature enabled

### DISCUSSION THREAD

**Discussion by johnnynotsolucky** [leptos-rs/leptos#1960] [Q&A]
I'm attempting to add islands functionality to my toy app, but have hit a couple of (hopefully simple) snags.

An island which renders a `<Outlet />` causes a warning and an error. The following code produces these logs:

```
You must call use_router() within a <Router/> component Location { file: ".cargo/registry/src/index.crates.io-6f17d22bba15001f/leptos_router-0.5.2/src/hooks.rs", line: 109, col: 53 } [test-leptos.js:487:9](http://127.0.0.1:3000/pkg/test-leptos.js)
```

and

```
Uncaught (in promise) RuntimeError: unreachable executed
    _island_ProvideState http://127.0.0.1:3000/pkg/test-leptos.js:233
    <anonymous> http://127.0.0.1:3000/:23
    promise callback* http://127.0.0.1:3000/:20
    promise callback* http://127.0.0.1:3000/:19
    requestIdleCallback handler*idle http://127.0.0.1:3000/:12
    <anonymous> http://127.0.0.1:3000/:17
[test-leptos.wasm:1403181:1](http://127.0.0.1:3000/pkg/test-leptos.wasm)
``` 

```rust
#[component]
pub fn App() -> impl IntoView {
    view! {
        <Router>
            <Routes>
                <Route path="/" view=ProvideState>
                    <Route path="/" view=SomeComponent />
                </Route>
            </Routes>
        </Router>
    }
}

#[island]
fn ProvideState() -> impl IntoView {
    view! {
        <Outlet />
    }
}

#[component]
fn SomeComponent() -> impl IntoView {
    view! {<p>Some Component</p>}
}
```

Its easy enough to resolve that by providing a closure to the Route's view prop instead:

```rust
#[derive(Debug, Clone)]
struct AppState;

#[component]
pub fn App() -> impl IntoView {
    view! {
        <Router>
            <Routes>
                <Route path="/" view=move || {
                    view! {
                        <ProvideState />
                        <Outlet />
                    }
                }>
                    <Route path="/" view=SomeComponent />
                </Route>
            </Routes>
        </Router>
    }
}

#[island]
fn ProvideState() -> impl IntoView {
    provide_context(AppState {});
    view! {}
}

#[component]
fn SomeComponent() -> impl IntoView {
    view! {<p>Some Component</p>}
}
```

However, now when I convert `SomeComponent` into an island and attempt to `expect_context`, a similar error as before occurs:

```
Uncaught (in promise) RuntimeError: unreachable executed
    _island_SomeComponent http://127.0.0.1:3000/pkg/test-leptos.js:211
    <anonymous> http://127.0.0.1:3000/:23
    promise callback* http://127.0.0.1:3000/:20
    promise callback* http://127.0.0.1:3000/:19
    requestIdleCallback handler*idle http://127.0.0.1:3000/:12
    <anonymous> http://127.0.0.1:3000/:17
[test-leptos.wasm:982261:1](http://127.0.0.1:3000/pkg/test-leptos.wasm)
```

```rust
#[island]
fn SomeComponent() -> impl IntoView {
    let app_state = expect_context::<AppState>();
    view! {<p>Some Component</p>}
}
```

But if I `use_context` instead and map the `Option`, it works correctly:

```rust
#[island]
fn SomeComponent() -> impl IntoView {
    let app_state = use_context::<AppState>();
    
    app_state.map(|state| {
        view! {<p>has state</p>}
    })
}
```

Is this expected behavior or am I going about islands in the wrong way?

My second question is around how data is loaded server-side in components. Before the islands feature was enabled I could use a resource to fetch data and wait for it with `<Await />`. That still works, but I noted [here](https://leptos-rs.github.io/leptos/islands.html#overview):

> Reduced #[server]/create_resource/Suspense boilerplate for loading server data.

So I was wondering, what is the recommended way to go about loading async data inside a component where currently I would have something like 

```rust
#[component]
fn User() -> IntoView {
	view! {
		<Await
			future=get_user
			let:user
		>
			// ...
		</Await>
	}
}
```

**Comment by gbj** [comment]
Taking this piece by piece:

**Using `<Outlet/>` in an `#[island]`**

Most of the `leptos_router` APIs are not available in the client in islands mode at present, including an `<Outlet/>` in an island: hence the first two errors.

**Passing context between siblings that are both islands**

This is kind of interesting. Context passes parent to child, but the "parents" in this case are not actually components -- rather they're the reactive ownership tree. So I guess you can pass context between siblings, because providing context in `ProvideState` actually provides it in the effect that creates both `ProvideState` and `SomeComponent`.

I'm not sure why `expect_context()` and `use_context().map()` would behave differently here; `expect_context()` is literally `use_context().unwrap()`. Are you sure the context is actually working in the `.map()` version, and not just being `None` and rendering nothing?

**Data fetching**

> My second question is around how data is loaded server-side in components. Before the islands feature was enabled I could use a resource to fetch data and wait for it with `<Await />`. That still works...

You should still just do that. The point about reduced boilerplate was that you *can* safely run server-only code in non-island `#[component]`s in islands mode, because they never execute on the client. So if you have some synchronous server work (like reading from a file with `std::fs`) you can do that in a component body. You might be able to block in place on a `Future` as well, I haven't actually tried this or thought it through too much!

But the same mental model of Suspense/Transition/Await still works identically in islands mode, so you don't need to change what you're already used to doing, and it should work just the same and be just as efficient. (Note also that HTML streaming with Suspense still works with islands and is still great.)

I'm uncertain about directly supporting `async` components even for server-only components in islands mode right now, because of the creation of data-loading waterfalls; if you could `.await` in a `#[component]` body that might be different but developing the best patterns for this (and whether it's actually a good idea) is still a very open question more broadly.

**Reply by bioinformatist** [reply]
This is REALLY an useful response. I just stucked on this data fetching topic these days. Thanks for reply and such a good question @johnnynotsolucky !

**Comment by johnnynotsolucky** [comment]
@gbj Thank you for walking through this, that makes perfect sense. 

> I'm not sure why expect_context() and use_context().map() would behave differently here; expect_context() is literally use_context().unwrap(). Are you sure the context is actually working in the .map() version, and not just being None and rendering nothing?

I tested with a match statement to confirm. I thought that maybe initially on the frontend it would be `None` before hydrating, but it doesn't seem to be the case:

```rust
#[island]
fn SomeComponent() -> impl IntoView {
	// let app_state = expect_context::<AppState>();
	// let app_state = use_context::<AppState>().unwrap();

    let app_state = use_context::<AppState>();

    match app_state {
        Some(state) => {
            log::warn!("STATE");
            view! {<p>has state</p>}.into_view()
        },
        None => {
            log::warn!("NO STATE");
            view!{<p>no state</p>}.into_view()
        }
    }
}
```

The only logs I see are for `"STATE"` and they appear on the server, not the client.

**Comment by jaroel** [comment]
I ran into the `Uncaught (in promise) RuntimeError: unreachable executed` too.
Turned out that I needed to have the `provide_context`-island wrap the `expect_context`-island.

ie,

```rust
#[component]
pub fn Works() -> impl IntoView {
    view! {
        <Player1>
            <Controls src="https://stream.dinxperfm.nl/1".to_string()/>
        </Player1>
    }
}

#[component]
pub fn DoesNotWork() -> impl IntoView {
    view! {
        <Player2/>
        <Controls src="https://stream.dinxperfm.nl/1".to_string()/>
    }
}

#[island]
pub fn Player1(children: Children) -> impl IntoView {
    let player_src: RwSignal<Option<String>> = RwSignal::new(None);
    provide_context(player_src);

    view! {
        <audio src=player_src></audio>
        {children()}
    }
}

#[island]
pub fn Player2() -> impl IntoView {
    let player_src: RwSignal<Option<String>> = RwSignal::new(None);
    provide_context(player_src);

    view! { <audio src=player_src></audio> }
}


#[island]
pub fn Controls(src: String) -> impl IntoView {
    let player_src = expect_context::<RwSignal<Option<String>>>();

    view! {
        <button
            on:click=move |_| {
                player_src.set(Some(src.clone()));
            }>
        "Play!"
        <button>
    }
}


```
